### ISSUE TITLE
Option to connect directly to each firewall instead of using Panorama as proxy

### ISSUE THREAD

**Issue by alexortize** [cdot65/pan-os-upgrade#101]
**Is your feature request related to a problem? Please describe.**
No but a potential one if too many calls to Panorama API endpoint.

**Describe the solution you'd like**
Provide a way to bypass default behaviour of proxying via Panorama

**Describe alternatives you've considered**
None

**Additional context**
We had issues when using upgrade assurance module in Ansible when proxying connection via Panorama

**Comment by cdot65** [comment]
Hello @alexortize, the `firewall` subcommand targets the firewalls directly, does that solve your need?

**Comment by alexortize** [comment]
I was more interested on direct connection when doing batch and HA
upgrades.

**Comment by cdot65** [comment]
Understood. A couple hurdles stand in the way and I'd be interested to getting your perspective:

## Inventory

Panorama provides a source of inventory as well as the connection to the devices. If we decide to target the firewalls directly, then we need to find a way of gathering a list of the devices and their IP addresses.

Since this is already captured when we run the `batch` or `inventory` commands, I'll assume that we can continue to keep this logic intact, but it would require a Panorama to be within the environment to derive our list from, is that okay?

Workflow would be like this: target Panorama to get a list of the devices, then use the IP address returned to form direct connections to the firewalls.

## Authentication

In order to pull this off, we would need to ensure that the authentication credentials work for both Panorama and every firewall. This may not be an issue for some environments, but it will pose challenges for others. Do you see any challenges with this requirement in your setup?

Multi-threading will prevent us from being able to prompt for unique username/password combinations across multiple firewalls.

**Comment by alexortize** [comment]
That’s perfect!

**Comment by alexortize** [comment]
I agree with you that this would work only for environment where same creds
work for both panorama and the firewalls.

**Comment by markusweb** [comment]
there are two points:

1) creating the inventory is ok to connect to panorama.
2) then why proxying the commands through panorama. why not have the option to connect to each firewall and do a upgrade firewall ... and provide a user/password for firewall (in some cases we have  a panorama and a firewall user)
would that not be faster to not use a "proxy command" but go directly on the firewall?

maybe provide a command switch to either have the batch update through panorama (proxied commands) or get the inventory from panorama but then do a direct firewall update with login to firewall mgmt ip?

Markus
